tests/appsec/waf/test_addresses.py

```diff
@@ -409,13 +411,127 @@ def test_main(self):
+            "/graphql",
+            headers={"Content-Type": "application/json"},
+            data=json.dumps(
+                {
+                    "query": "query getUserByName($name: String) { userByName(name: $name) { id name }}",
+                    "variables": {"name": "foo"},
+                    "operationName": "getUserByName",
+                }
+            ),
+        )
+
+    def test_request_no_attack(self):
+        """Verify that no AppSec event was reported"""
+
+        assert self.r_no_attack.status_code == 200  # There is no attack here!
+        interfaces.library.assert_no_appsec_event(self.r_no_attack)
+
+    def setup_request_monitor_attack(self):
+        """Set up a request with a resolver-targeted attack"""
+
+        self.r_attack = weblog.post(
+            "/graphql",
+            headers={"Content-Type": "application/json"},
+            data=json.dumps(
+                {
+                    "query": "query getUserByName($name: String) { userByName(name: $name) { id name }}",
+                    "variables": {"name": "testattack"},
+                    "operationName": "getUserByName",
+                }
+            ),
+        )
+
+    def test_request_monitor_attack(self):
+        """Verify that the request triggered a resolver attack event"""
+
+        assert self.r_attack.status_code == 200  # This attack is never blocking
+
+        failures = []
+
+        try:
+            interfaces.library.assert_waf_attack(
+                self.r_attack,
+                rule="monitor-resolvers",
+                key_path=["userByName", "name"],
+                value="testattack",
+                full_trace=True,
+            )
+        except ValueError as e:
+            failures.append(e)
+
+        try:
+            interfaces.library.assert_waf_attack(
+                self.r_attack,
+                rule="monitor-all-resolvers",
+                key_path=["userByName", "0", "name"],
+                value="testattack",
+                full_trace=True,
+            )
+        except ValueError as e:
+            failures.append(e)
+
+        # At least one of the two assertions should have passed...
+        if len(failures) >= 2:
+            raise ExceptionGroup(f"At least one rule should have triggered\n- {failures[0]}\n- {failures[1]}", failures)
+
+    def setup_request_monitor_attack_directive(self):
+        """Set up a request with a directive-targeted attack"""
+
+        self.r_attack = weblog.post(
+            "/graphql",
+            headers={"Content-Type": "application/json"},
+            data=json.dumps(
+                {
+                    "query": 'query getUserByName($name: String) { userByName(name: $name) @case(format: "testattack") { id name }}',
+                    "variables": {"name": "test"},
+                    "operationName": "getUserByName",
+                }
+            ),
+        )
+
+    def test_request_monitor_attack_directive(self):
```

Comment:
Suggestion, this method is very similar to test_request_monitor_attack, perhaps we can use a generic method and call it from the both methods

Reply:
extracted the logic to new method